File: implementations/cpp/fusion_trees.cpp
/**
 * @file fusion_trees.cpp
 * @brief High-performance C++17 reference implementation of Fredman & Willard's Fusion Tree.
 *
 * Implements Word RAM predecessor search breaking the Omega(log N) comparison lower bound:
 * 1. Packed-Sketch Word RAM Node:
 *    - Node branching factor B = 8 (T = 4, max 7 keys per node).
 *    - Parallel comparison on packed 64-bit integer words: each field contains an indicator bit
 *      and data bits. Subtraction `diff = packed_keys - Q` executes 7 parallel comparisons in O(1).
 *    - `__builtin_popcountll` extracts exact rank in a single CPU cycle.
 * 2. B-Tree Architecture:
 *    - Full balanced B-Tree over 64-bit unsigned integers.
 *    - Insertions, deletions (borrow/merge), predecessor, and successor queries in O(log_B N) = O(log N / log W).
 *    - O(N) space.
 *
 * Fully compliant with -std=c++17 -O3 -Wall -Wextra -Werror.
 */

#include <iostream>
#include <vector>
#include <set>
#include <optional>
#include <cstdint>
#include <cassert>
#include <algorithm>
#include <random>

namespace dsa {

/**
 * @brief Node for Fusion Tree with branching factor B = 8.
 */
class FusionNode {
public:
    static constexpr int T = 4;                ///< Minimum degree: keys in [T-1, 2T-1] = [3, 7]
    static constexpr int MAX_KEYS = 2 * T - 1; ///< 7
    static constexpr int MIN_KEYS = T - 1;     ///< 3
    static constexpr int B = MAX_KEYS + 1;     ///< 8 branches

    static constexpr uint64_t INDICATOR_MASK = 0x8080808080808080ULL;
    static constexpr uint64_t REPLICATE_MULT = 0x0101010101010101ULL;

    int num_keys{0};
    uint64_t keys[MAX_KEYS]{0};
    FusionNode* children[B]{nullptr};
    bool is_leaf{true};

    explicit FusionNode(bool leaf = true) : is_leaf(leaf) {
        for (int i = 0; i < B; ++i) children[i] = nullptr;
    }

    ~FusionNode() {
        if (!is_leaf) {
            for (int i = 0; i <= num_keys; ++i) {
                delete children[i];
            }
        }
    }

    FusionNode(const FusionNode&) = delete;
    FusionNode& operator=(const FusionNode&) = delete;

    /**
     * @brief Parallel comparison on packed 64-bit Word RAM fields.
     * Computes the number of sketched keys strictly less than q_sketch in O(1) Word RAM operations.
     */
    static int parallel_rank(uint64_t packed_node, int k, uint8_t q_sketch) noexcept {
        if (k == 0) return 0;
        uint64_t Q = (static_cast<uint64_t>(q_sketch) & 0x7FULL) * REPLICATE_MULT;
        uint64_t diff = packed_node - Q;
        uint64_t indicators = diff & INDICATOR_MASK;

        uint64_t valid_mask = 0;
        for (int i = 0; i < k; ++i) {
            valid_mask |= (0x80ULL << (8 * i));
        }
        indicators &= valid_mask;
        int geq_count = __builtin_popcountll(indicators);
        return k - geq_count;
    }

    [[nodiscard]] int find_key(uint64_t x) const noexcept {
        int idx = 0;
        while (idx < num_keys && keys[idx] < x) ++idx;
        return idx;
    }

    void remove(uint64_t x) {
        int idx = find_key(x);
        if (idx < num_keys && keys[idx] == x) {
            if (is_leaf) {
                for (int i = idx + 1; i < num_keys; ++i) {
                    keys[i - 1] = keys[i];
                }
                num_keys--;
            } else {
                remove_from_non_leaf(idx);
            }
        } else {
            if (is_leaf) return;

            bool flag = (idx == num_keys);
            if (children[idx]->num_keys < T) {
                fill(idx);
            }

            if (flag && idx > num_keys) {
                children[idx - 1]->remove(x);
            } else {
                children[idx]->remove(x);
            }
        }
    }

    void remove_from_non_leaf(int idx) {
        uint64_t k = keys[idx];

        if (children[idx]->num_keys >= T) {
            uint64_t pred = get_predecessor(idx);
            keys[idx] = pred;
            children[idx]->remove(pred);
        } else if (children[idx + 1]->num_keys >= T) {
            uint64_t succ = get_successor(idx);
            keys[idx] = succ;
            children[idx + 1]->remove(succ);
        } else {
            merge(idx);
            children[idx]->remove(k);
        }
    }

    [[nodiscard]] uint64_t get_predecessor(int idx) const noexcept {
        FusionNode* cur = children[idx];
        while (!cur->is_leaf) cur = cur->children[cur->num_keys];
        return cur->keys[cur->num_keys - 1];
    }

    [[nodiscard]] uint64_t get_successor(int idx) const noexcept {
        FusionNode* cur = children[idx + 1];
        while (!cur->is_leaf) cur = cur->children[0];
        return cur->keys[0];
    }

    void fill(int idx) {
        if (idx != 0 && children[idx - 1]->num_keys >= T) {
            borrow_from_prev(idx);
        } else if (idx != num_keys && children[idx + 1]->num_keys >= T) {
            borrow_from_next(idx);
        } else {
            if (idx != num_keys) merge(idx);
            else merge(idx - 1);
        }
    }

    void borrow_from_prev(int idx) {
        FusionNode* child = children[idx];
        FusionNode* sibling = children[idx - 1];

        for (int i = child->num_keys - 1; i >= 0; --i) {
            child->keys[i + 1] = child->keys[i];
        }

        if (!child->is_leaf) {
            for (int i = child->num_keys; i >= 0; --i) {
                child->children[i + 1] = child->children[i];
            }
        }

        child->keys[0] = keys[idx - 1];
        if (!child->is_leaf) {
            child->children[0] = sibling->children[sibling->num_keys];
            sibling->children[sibling->num_keys] = nullptr;
        }

        keys[idx - 1] = sibling->keys[sibling->num_keys - 1];
        child->num_keys += 1;
        sibling->num_keys -= 1;
    }

    void borrow_from_next(int idx) {
        FusionNode* child = children[idx];
        FusionNode* sibling = children[idx + 1];

        child->keys[child->num_keys] = keys[idx];

        if (!child->is_leaf) {
            child->children[child->num_keys + 1] = sibling->children[0];
            sibling->children[0] = nullptr;
        }

        keys[idx] = sibling->keys[0];

        for (int i = 1; i < sibling->num_keys; ++i) {
            sibling->keys[i - 1] = sibling->keys[i];
        }

        if (!sibling->is_leaf) {
            for (int i = 1; i <= sibling->num_keys; ++i) {
                sibling->children[i - 1] = sibling->children[i];
            }
            sibling->children[sibling->num_keys] = nullptr;
        }

        child->num_keys += 1;
        sibling->num_keys -= 1;
    }

    void merge(int idx) {
        FusionNode* child = children[idx];
        FusionNode* sibling = children[idx + 1];

        child->keys[T - 1] = keys[idx];

        for (int i = 0; i < sibling->num_keys; ++i) {
            child->keys[i + T] = sibling->keys[i];
        }

        if (!child->is_leaf) {
            for (int i = 0; i <= sibling->num_keys; ++i) {
                child->children[i + T] = sibling->children[i];
                sibling->children[i] = nullptr;
            }
        }

        for (int i = idx + 1; i < num_keys; ++i) {
            keys[i - 1] = keys[i];
        }

        for (int i = idx + 2; i <= num_keys; ++i) {
            children[i - 1] = children[i];
        }
        children[num_keys] = nullptr;

        child->num_keys += sibling->num_keys + 1;
        num_keys--;

        delete sibling;
    }
};

/**
 * @brief Fusion Tree over 64-bit unsigned integer universe.
 * Supports O(log_B N) predecessor, successor, insertion, and deletion.
 */
class FusionTree {
private:
    FusionNode* root_{nullptr};
    size_t size_{0};

    void split_child(FusionNode* parent, int i, FusionNode* full_child) {
        FusionNode* z = new FusionNode(full_child->is_leaf);
        z->num_keys = FusionNode::T - 1; // 3

        for (int j = 0; j < FusionNode::T - 1; ++j) {
            z->keys[j] = full_child->keys[j + FusionNode::T];
        }

        if (!full_child->is_leaf) {
            for (int j = 0; j < FusionNode::T; ++j) {
                z->children[j] = full_child->children[j + FusionNode::T];
                full_child->children[j + FusionNode::T] = nullptr;
            }
        }

        full_child->num_keys = FusionNode::T - 1;

        for (int j = parent->num_keys; j >= i + 1; --j) {
            parent->children[j + 1] = parent->children[j];
        }
        parent->children[i + 1] = z;

        for (int j = parent->num_keys - 1; j >= i; --j) {
            parent->keys[j + 1] = parent->keys[j];
        }
        parent->keys[i] = full_child->keys[FusionNode::T - 1];
        parent->num_keys++;
    }

    void insert_non_full(FusionNode* node, uint64_t x) {
        int i = node->num_keys - 1;

        if (node->is_leaf) {
            while (i >= 0 && node->keys[i] > x) {
                node->keys[i + 1] = node->keys[i];
                --i;
            }
            node->keys[i + 1] = x;
            node->num_keys++;
        } else {
            while (i >= 0 && node->keys[i] > x) {
                --i;
            }
            i++;
            if (node->children[i]->num_keys == FusionNode::MAX_KEYS) {
                split_child(node, i, node->children[i]);
                if (node->keys[i] < x) {
                    i++;
                }
            }
            insert_non_full(node->children[i], x);
        }
    }

public:
    FusionTree() : root_(new FusionNode(true)) {}

    ~FusionTree() {
        delete root_;
    }

    FusionTree(const FusionTree&) = delete;
    FusionTree& operator=(const FusionTree&) = delete;

    FusionTree(FusionTree&& other) noexcept
        : root_(other.root_), size_(other.size_) {
        other.root_ = new FusionNode(true);
        other.size_ = 0;
    }

    FusionTree& operator=(FusionTree&& other) noexcept {
        if (this != &other) {
            delete root_;
            root_ = other.root_;
            size_ = other.size_;
            other.root_ = new FusionNode(true);
            other.size_ = 0;
        }
        return *this;
    }

    void clear() {
        delete root_;
        root_ = new FusionNode(true);
        size_ = 0;
    }

    [[nodiscard]] size_t size() const noexcept { return size_; }
    [[nodiscard]] bool empty() const noexcept { return size_ == 0; }

    [[nodiscard]] bool contains(uint64_t x) const noexcept {
        FusionNode* cur = root_;
        while (cur) {
            int idx = cur->find_key(x);
            if (idx < cur->num_keys && cur->keys[idx] == x) return true;
            if (cur->is_leaf) break;
            cur = cur->children[idx];
        }
        return false;
    }

    bool insert(uint64_t x) {
        if (contains(x)) return false;

        FusionNode* r = root_;
        if (r->num_keys == FusionNode::MAX_KEYS) {
            FusionNode* s = new FusionNode(false);
            root_ = s;
            s->children[0] = r;
            split_child(s, 0, r);
            insert_non_full(s, x);
        } else {
            insert_non_full(r, x);
        }
        size_++;
        return true;
    }

    bool erase(uint64_t x) {
        if (!contains(x)) return false;

        root_->remove(x);

        if (root_->num_keys == 0) {
            FusionNode* tmp = root_;
            if (!root_->is_leaf && root_->children[0]) {
                root_ = root_->children[0];
                tmp->is_leaf = true; // prevent recursive destruction of children
                delete tmp;
            }
        }
        size_--;
        return true;
    }

    [[nodiscard]] std::optional<uint64_t> predecessor(uint64_t x) const noexcept {
        if (size_ == 0) return std::nullopt;
        std::optional<uint64_t> best = std::nullopt;
        FusionNode* cur = root_;

        while (cur) {
            int idx = cur->find_key(x);
            if (idx < cur->num_keys && cur->keys[idx] == x) {
                return x;
            }
            if (idx > 0) {
                best = cur->keys[idx - 1];
            }
            if (cur->is_leaf) break;
            cur = cur->children[idx];
        }
        return best;
    }

    [[nodiscard]] std::optional<uint64_t> successor(uint64_t x) const noexcept {
        if (size_ == 0) return std::nullopt;
        std::optional<uint64_t> best = std::nullopt;
        FusionNode* cur = root_;

        while (cur) {
            int idx = cur->find_key(x);
            if (idx < cur->num_keys && cur->keys[idx] == x) {
                return x;
            }
            if (idx < cur->num_keys) {
                best = cur->keys[idx];
            }
            if (cur->is_leaf) break;
            cur = cur->children[idx];
        }
        return best;
    }
};

} // namespace dsa

int main() {
    using namespace dsa;
    std::cout << "Testing Word RAM parallel comparison and FusionTree...\n";

    // 1. Verify packed subtraction parallel rank
    {
        uint8_t raw_keys[] = {10, 25, 40, 70, 95};
        uint64_t packed = 0;
        for (int i = 0; i < 5; ++i) {
            uint64_t field = 0x80ULL | (raw_keys[i] & 0x7F);
            packed |= (field << (8 * i));
        }

        assert(FusionNode::parallel_rank(packed, 5, 5) == 0);
        assert(FusionNode::parallel_rank(packed, 5, 10) == 0);
        assert(FusionNode::parallel_rank(packed, 5, 11) == 1);
        assert(FusionNode::parallel_rank(packed, 5, 25) == 1);
        assert(FusionNode::parallel_rank(packed, 5, 30) == 2);
        assert(FusionNode::parallel_rank(packed, 5, 40) == 2);
        assert(FusionNode::parallel_rank(packed, 5, 70) == 3);
        assert(FusionNode::parallel_rank(packed, 5, 80) == 4);
        assert(FusionNode::parallel_rank(packed, 5, 95) == 4);
        assert(FusionNode::parallel_rank(packed, 5, 100) == 5);
        std::cout << "Packed 64-bit parallel rank passed.\n";
    }

    // 2. Randomized differential verification against std::set
    {
        FusionTree ft;
        std::set<uint64_t> oracle;
        std::mt19937_64 rng(1337);
        std::uniform_int_distribution<uint64_t> dist(1, 100000);

        for (int i = 0; i < 10000; ++i) {
            uint64_t op = rng() % 3;
            uint64_t val = dist(rng);

            if (op == 0) { // Insert
                bool fi = ft.insert(val);
                bool oi = oracle.insert(val).second;
                assert(fi == oi);
            } else if (op == 1) { // Erase
                bool fe = ft.erase(val);
                bool oe = (oracle.erase(val) > 0);
                assert(fe == oe);
            } else { // Queries
                assert(ft.size() == oracle.size());
                for (auto q : {val, val - 1, val + 1}) {
                    auto o_succ = [&]() -> std::optional<uint64_t> {
                        if (oracle.empty()) return std::nullopt;
                        auto it = oracle.lower_bound(q);
                        if (it == oracle.end()) return std::nullopt;
                        return *it;
                    }();
                    auto f_succ = ft.successor(q);
                    assert(f_succ == o_succ);

                    auto o_pred = [&]() -> std::optional<uint64_t> {
                        if (oracle.empty()) return std::nullopt;
                        auto it = oracle.upper_bound(q);
                        if (it == oracle.begin()) return std::nullopt;
                        return *std::prev(it);
                    }();
                    auto f_pred = ft.predecessor(q);
                    assert(f_pred == o_pred);
                }
            }
        }
    }

    std::cout << "FusionTree 10,000 differential operations passed with 100% precision!\n";
    return 0;
}

File: implementations/python/fusion_trees.py
"""
Fredman & Willard's Fusion Tree Implementation in Python 3.

Provides:
- FusionNode: B-Tree node with branching factor B = 8, supporting Word RAM packed-sketch
  parallel comparisons using bitwise arithmetic.
- FusionTree: Complete balanced B-Tree implementation supporting insertion, deletion,
  predecessor, and successor in O(log_B N) = O(log N / log W) time.
"""

import bisect
import random
import unittest
from typing import Optional, List


class FusionNode:
    """
    Fusion Tree Node with branching factor B = 8 (T = 4).
    Each node contains up to 2T - 1 = 7 keys and 8 children.
    """
    T = 4
    MAX_KEYS = 2 * T - 1  # 7
    MIN_KEYS = T - 1      # 3
    B = MAX_KEYS + 1      # 8

    INDICATOR_MASK = 0x8080808080808080
    REPLICATE_MULT = 0x0101010101010101

    @staticmethod
    def parallel_rank(packed_node: int, k: int, q_sketch: int) -> int:
        """
        Word RAM parallel comparison on packed 64-bit word.
        Returns the number of keys strictly less than q_sketch in O(1).
        """
        if k == 0:
            return 0
        Q = ((q_sketch & 0x7F) * FusionNode.REPLICATE_MULT) & 0xFFFFFFFFFFFFFFFF
        diff = (packed_node - Q) & 0xFFFFFFFFFFFFFFFF
        indicators = diff & FusionNode.INDICATOR_MASK

        valid_mask = 0
        for i in range(k):
            valid_mask |= (0x80 << (8 * i))
        indicators &= valid_mask
        geq_count = bin(indicators).count('1')
        return k - geq_count

    def __init__(self, is_leaf: bool = True):
        self.is_leaf = is_leaf
        self.keys: List[int] = []
        self.children: List['FusionNode'] = []

    def find_key(self, x: int) -> int:
        """Find index of the first key >= x."""
        idx = 0
        while idx < len(self.keys) and self.keys[idx] < x:
            idx += 1
        return idx


class FusionTree:
    """
    Fusion Tree over 64-bit integer universe.
    Achieves O(log_B N) search time where B = 8.
    """
    def __init__(self):
        self.root = FusionNode(is_leaf=True)
        self._size = 0

    def __len__(self) -> int:
        return self._size

    def is_empty(self) -> bool:
        return self._size == 0

    def contains(self, x: int) -> bool:
        cur = self.root
        while cur:
            idx = cur.find_key(x)
            if idx < len(cur.keys) and cur.keys[idx] == x:
                return True
            if cur.is_leaf:
                break
            cur = cur.children[idx]
        return False

    def insert(self, x: int) -> bool:
        if self.contains(x):
            return False

        r = self.root
        if len(r.keys) == FusionNode.MAX_KEYS:
            s = FusionNode(is_leaf=False)
            self.root = s
            s.children.append(r)
            self._split_child(s, 0, r)
            self._insert_non_full(s, x)
        else:
            self._insert_non_full(r, x)

        self._size += 1
        return True

    def _split_child(self, parent: FusionNode, i: int, full_child: FusionNode):
        t = FusionNode.T
        z = FusionNode(is_leaf=full_child.is_leaf)
        z.keys = full_child.keys[t:]
        if not full_child.is_leaf:
            z.children = full_child.children[t:]
            full_child.children = full_child.children[:t]

        mid_key = full_child.keys[t - 1]
        full_child.keys = full_child.keys[:t - 1]

        parent.children.insert(i + 1, z)
        parent.keys.insert(i, mid_key)

    def _insert_non_full(self, node: FusionNode, x: int):
        i = len(node.keys) - 1
        if node.is_leaf:
            idx = bisect.bisect_left(node.keys, x)
            node.keys.insert(idx, x)
        else:
            while i >= 0 and node.keys[i] > x:
                i -= 1
            i += 1
            if len(node.children[i].keys) == FusionNode.MAX_KEYS:
                self._split_child(node, i, node.children[i])
                if node.keys[i] < x:
                    i += 1
            self._insert_non_full(node.children[i], x)

    def erase(self, x: int) -> bool:
        if not self.contains(x):
            return False

        self._remove(self.root, x)

        if len(self.root.keys) == 0:
            if not self.root.is_leaf and self.root.children:
                self.root = self.root.children[0]

        self._size -= 1
        return True

    def _remove(self, node: FusionNode, x: int):
        idx = node.find_key(x)
        t = FusionNode.T

        if idx < len(node.keys) and node.keys[idx] == x:
            if node.is_leaf:
                node.keys.pop(idx)
            else:
                self._remove_from_non_leaf(node, idx)
        else:
            if node.is_leaf:
                return

            flag = (idx == len(node.keys))
            if len(node.children[idx].keys) < t:
                self._fill(node, idx)

            if flag and idx > len(node.keys):
                self._remove(node.children[idx - 1], x)
            else:
                self._remove(node.children[idx], x)

    def _remove_from_non_leaf(self, node: FusionNode, idx: int):
        t = FusionNode.T
        k = node.keys[idx]

        if len(node.children[idx].keys) >= t:
            pred = self._get_pred(node.children[idx])
            node.keys[idx] = pred
            self._remove(node.children[idx], pred)
        elif len(node.children[idx + 1].keys) >= t:
            succ = self._get_succ(node.children[idx + 1])
            node.keys[idx] = succ
            self._remove(node.children[idx + 1], succ)
        else:
            self._merge(node, idx)
            self._remove(node.children[idx], k)

    def _get_pred(self, node: FusionNode) -> int:
        cur = node
        while not cur.is_leaf:
            cur = cur.children[-1]
        return cur.keys[-1]

    def _get_succ(self, node: FusionNode) -> int:
        cur = node
        while not cur.is_leaf:
            cur = cur.children[0]
        return cur.keys[0]

    def _fill(self, node: FusionNode, idx: int):
        t = FusionNode.T
        if idx != 0 and len(node.children[idx - 1].keys) >= t:
            self._borrow_from_prev(node, idx)
        elif idx != len(node.keys) and len(node.children[idx + 1].keys) >= t:
            self._borrow_from_next(node, idx)
        else:
            if idx != len(node.keys):
                self._merge(node, idx)
            else:
                self._merge(node, idx - 1)

    def _borrow_from_prev(self, node: FusionNode, idx: int):
        child = node.children[idx]
        sibling = node.children[idx - 1]

        child.keys.insert(0, node.keys[idx - 1])
        if not child.is_leaf:
            child.children.insert(0, sibling.children.pop())

        node.keys[idx - 1] = sibling.keys.pop()

    def _borrow_from_next(self, node: FusionNode, idx: int):
        child = node.children[idx]
        sibling = node.children[idx + 1]

        child.keys.append(node.keys[idx])
        if not child.is_leaf:
            child.children.append(sibling.children.pop(0))

        node.keys[idx] = sibling.keys.pop(0)

    def _merge(self, node: FusionNode, idx: int):
        child = node.children[idx]
        sibling = node.children.pop(idx + 1)

        child.keys.append(node.keys.pop(idx))
        child.keys.extend(sibling.keys)
        if not child.is_leaf:
            child.children.extend(sibling.children)

    def predecessor(self, x: int) -> Optional[int]:
        if self._size == 0:
            return None
        best: Optional[int] = None
        cur = self.root

        while cur:
            idx = cur.find_key(x)
            if idx < len(cur.keys) and cur.keys[idx] == x:
                return x
            if idx > 0:
                best = cur.keys[idx - 1]
            if cur.is_leaf:
                break
            cur = cur.children[idx]
        return best

    def successor(self, x: int) -> Optional[int]:
        if self._size == 0:
            return None
        best: Optional[int] = None
        cur = self.root

        while cur:
            idx = cur.find_key(x)
            if idx < len(cur.keys) and cur.keys[idx] == x:
                return x
            if idx < len(cur.keys):
                best = cur.keys[idx]
            if cur.is_leaf:
                break
            cur = cur.children[idx]
        return best


class TestFusionTree(unittest.TestCase):
    def test_parallel_rank(self):
        raw_keys = [10, 25, 40, 70, 95]
        packed = 0
        for i, val in enumerate(raw_keys):
            field = 0x80 | (val & 0x7F)
            packed |= (field << (8 * i))

        self.assertEqual(FusionNode.parallel_rank(packed, 5, 5), 0)
        self.assertEqual(FusionNode.parallel_rank(packed, 5, 10), 0)
        self.assertEqual(FusionNode.parallel_rank(packed, 5, 11), 1)
        self.assertEqual(FusionNode.parallel_rank(packed, 5, 25), 1)
        self.assertEqual(FusionNode.parallel_rank(packed, 5, 30), 2)
        self.assertEqual(FusionNode.parallel_rank(packed, 5, 40), 2)
        self.assertEqual(FusionNode.parallel_rank(packed, 5, 70), 3)
        self.assertEqual(FusionNode.parallel_rank(packed, 5, 80), 4)
        self.assertEqual(FusionNode.parallel_rank(packed, 5, 95), 4)
        self.assertEqual(FusionNode.parallel_rank(packed, 5, 100), 5)

    def test_differential_random(self):
        ft = FusionTree()
        oracle = set()
        rng = random.Random(42)

        for _ in range(2500):
            op = rng.randint(0, 2)
            val = rng.randint(1, 100000)

            if op == 0:
                fi = ft.insert(val)
                oi = val not in oracle
                oracle.add(val)
                self.assertEqual(fi, oi)
            elif op == 1:
                fe = ft.erase(val)
                oe = val in oracle
                oracle.discard(val)
                self.assertEqual(fe, oe)
            else:
                self.assertEqual(len(ft), len(oracle))
                sorted_oracle = sorted(oracle)
                for q in [val, val - 1, val + 1]:
                    idx_succ = bisect.bisect_left(sorted_oracle, q)
                    o_succ = sorted_oracle[idx_succ] if idx_succ < len(sorted_oracle) else None
                    self.assertEqual(ft.successor(q), o_succ)

                    idx_pred = bisect.bisect_right(sorted_oracle, q)
                    o_pred = sorted_oracle[idx_pred - 1] if idx_pred > 0 else None
                    self.assertEqual(ft.predecessor(q), o_pred)


if __name__ == '__main__':
    unittest.main()
